Add num_workers and save_main_session flag to auto_model_refresh notebook

AnandInguva opened the pull request:
Fixes: https://github.com/apache/beam/issues/28773

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

AnandInguva commented:
R: @damccorm 

I used colab to push it to the branch. I am not sure why the spaces and everything got formatted. When I open the notebook locally, others cells with metadata are getting edited when I push it to the branch(I guess due to the IDE I use)

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,652 @@
+      "cell_type": "code",
+      "source": [
+        "# Authenticate to your Google Cloud account.\n",
+        "from google.colab import auth\n",
```
I found that this broke the `save_main_session=True` case because it got saved as part of the main session and then worker startup failed trying to do this import.

The fix was to make this cell:

```
# Authenticate to your Google Cloud account.
def auth_to_colab():
  from google.colab import auth
  auth.authenticate_user()

auth_to_colab()
```

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,652 @@
+        "## Use the TensorFlow model handler\n",
+        " This example uses `TFModelHandlerTensor` as the model handler and the `resnet_101` model trained on [ImageNet](https://www.image-net.org/).\n",
+        "\n",
+        " Download the model from [Google Cloud Storage](https://storage.googleapis.com/tensorflow/keras-applications/resnet/resnet101_weights_tf_dim_ordering_tf_kernels.h5) (link downloads the model), and place it in the directory that you want to use to update your model.\n",
```
I found that just downloading the model naively didn't work and I got an error about it being saved in an improper format. This can be demonstrated by using the same command our model handler uses for loading: `tf.keras.models.load_model('path/to/resnet101_weights_tf_dim_ordering_tf_kernels.h5')`

The fix I found was to load the model and then save it with keras. So:

```
model = tf.keras.applications.resnet.ResNet101()
model.save('path/to/resnet101_weights_tf_dim_ordering_tf_kernels.keras')
model = tf.keras.applications.resnet.ResNet152()
model.save('path/to/resnet152_weights_tf_dim_ordering_tf_kernels.keras')
```

(and then using the `.keras` extension elsewhere). I think the `.h5` files just contain the model weights

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,652 @@
+        "from typing import Tuple\n",
+        "\n",
+        "import apache_beam as beam\n",
+        "from apache_beam.examples.inference.tensorflow_imagenet_segmentation import PostProcessor\n",
```
We don't actually use this import

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,652 @@
+        "\n",
+        "import apache_beam as beam\n",
+        "from apache_beam.examples.inference.tensorflow_imagenet_segmentation import PostProcessor\n",
+        "from apache_beam.examples.inference.tensorflow_imagenet_segmentation import read_image\n",
```
Can we inline `read_image` instead of importing it from an example (which is liable to change)

damccorm reviewed (commented):
Thanks for following up on this quickly

AnandInguva commented:
Addressed comments.

`.h5` used to store entire model as well. I was loading model with the same link but may be in the updated version, something has changed? In the new TF version, they are referring `.h5` to be legacy and `.keras` was introduced. So its better to keep `.keras`

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -99,27 +99,14 @@
         "!pip install apache_beam[gcp]>=2.46.0 --quiet\n",
-        "!pip install tensorflow\n",
-        "!pip install tensorflow_hub"
+        "# !pip install tensorflow_hub"
```
Did you mean to change these locally?

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -132,8 +119,6 @@
         "\n",
         "import apache_beam as beam\n",
-        "from apache_beam.examples.inference.tensorflow_imagenet_segmentation import PostProcessor\n",
-        "from apache_beam.examples.inference.tensorflow_imagenet_segmentation import read_image\n",
```
We still need to inline this function somewhere (probably alongside the mapping function where we call it)

AnandInguva replied:
Can you explain it? I didn't understand `inline` this function?

damccorm replied:
Right now, read_image is still called, but its not defined anywhere. Originally we imported it, but that's fragile since the example could change/doesn't have rigorous API guarantees. We should define the function next to its usage instead

AnandInguva replied:
Yeah, we use `preprocess_image` and this change was also lost in the commit. I am running the notebook with updated changes.

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -224,29 +202,13 @@
-        "# To expedite the model update process, it's recommended to set num_workers>1.\n",
-        "# https://github.com/apache/beam/issues/28776\n",
-        "options.view_as(WorkerOptions).num_workers = 5"
+        "\n"
```
Did you mean to remove save_main_session/num_workers?

AnandInguva commented:
I opened the notebook with the option open the notebook using colab and it removed the code suggestions and previous commits. It gave the master ref. That was why the changes were not in the recent commit. I made changes now. PTAL

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,668 @@
+    {
+      "cell_type": "code",
+      "source": [
+        "model = tf.keras.applications.resnet.ResNet101()\n",
```
We should have something similar to get the resnet 152 model below

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,668 @@
+      "source": [
+        "3. Pass the images to the RunInference `PTransform`. RunInference takes `model_handler` and `model_metadata_pcoll` as input parameters.\n",
+        "  * `model_metadata_pcoll` is a side input `PCollection` to the RunInference `PTransform`. This side input is used to update the `model_uri` in the `model_handler` without needing to stop the Apache Beam pipeline\n",
+        "  * Use `WatchFilePattern` as side input to watch a `file_pattern` matching `.h5` files. In this case, the `file_pattern` is `'gs://BUCKET_NAME/*.h5'`.\n",
```
Should be `*.keras` here and below

damccorm commented:
> I opened the notebook with the option open the notebook using colab and it removed the code suggestions and previous commits. It gave the master ref. That was why the changes were not in the recent commit. I made changes now. PTAL

You can reference it with your repo/branch name FWIW - https://colab.sandbox.google.com/github/AnandInguva/beam/blob/auto_model_refresh/examples/notebooks/beam-ml/automatic_model_refresh.ipynb - I'd recommend running through it that way after pushing the branch but before opening PRs in general to check your logic

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,530 @@
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/AnandInguva/beam/blob/auto_model_refresh/examples/notebooks/beam-ml/automatic_model_refresh.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
```suggestion
        "<a href=\"https://colab.research.google.com/github/apache/beam/blob/master/examples/notebooks/beam-ml/automatic_model_refresh.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

damccorm commented on examples/notebooks/beam-ml/automatic_model_refresh.ipynb:
```diff
@@ -1,605 +1,530 @@
+        "model = tf.keras.applications.resnet.ResNet152()\n",
+        "model.save('resnet152_weights_tf_dim_ordering_tf_kernels.keras')\n",
+        "# Replace the `BUCKET_NAME` with the actual bucket name.\n",
+        "!gsutil cp resnet152_weights_tf_dim_ordering_tf_kernels.keras gs://anandinguva-test/<BUCKET_NAME>/resnet152_weights_tf_dim_ordering_tf_kernels.keras"
```
```suggestion
        "!gsutil cp resnet152_weights_tf_dim_ordering_tf_kernels.keras gs://<BUCKET_NAME>/resnet152_weights_tf_dim_ordering_tf_kernels.keras"
```

damccorm reviewed (commented):
Mostly LGTM, just a couple more comments

damccorm commented:
FYI, I'd recommend doing changes in a text editor or on GitHub to avoid a big diff from colab

damccorm reviewed (approved):
Thanks!
